eng/common/TestResources/New-TestResources.ps1

```diff
@@ -59,7 +59,10 @@ param (
     [string] $Environment = 'AzureCloud',
 
     [Parameter()]
-    [hashtable] $AdditionalParameters,
+    [hashtable] $ArmTemplateParameters,
```

Comment:
This is a breaking change. This script isn't just used by CIs, but by devs. Why this change?

Reply:
The script already removes variables that aren't declared as parameters from the ARM template.

Reply:
Agreed we should add a new parameter instead of replacing it. We can join them later in the script as needed.

Reply:
Do we even need to do that, though? The script already passes all parameters to the test-resources-post.ps1 script (if present), so why separate variables? At the very least, any parameters passed in should be merged to pass on through.

Reply:
I like moving a lot of the default parameters into the sub-config and I'm definitely open for allowing some tests to enable passing additional parameters in special cases. It is just a matter of where we merge them. I suggested merging the script so that we don't have to do a lot of inline scripting outside the script and we don't have to do much remapping of the sub-config to special case this.

Reply:
I don't think we're talking about the same thing. I'm saying the `AdditionalParameters` (i.e. not changing it to `ArmTemplateParameters`) and `EnvironmentVariables` don't need to be separate. Extraneous variables of any kind won't be passed to the ARM template. So why not just pass them all in as `AdditionalParameters`?

Reply:
I didn't realize this script was used by devs in other places, hence I thought I could control the breaking change. Thanks for pointing that out :) Can either of you provide more context on who/how people use this? I think the logic sharing is good, but at the same time having devs take a dependency on our provisioning scripts means we can basically never make breaking changes without some sort of non-standardized effort to notify people. I'm not sure this is a good situation to be in.

Re: AdditionalParameters, I think the name could be more descriptive. The parameters are only used for ARM templates, but the naming does not make that clear. Similarly, the type is different as AdditionalParameters/ArmTemplateParameters is an object. Someone looking at a tests.yml file may be confused trying to understand the difference between matrix variables (string type) defined in AdditionalPlatforms items, and parameters defined in a field named AdditionalParameters. If they are only intended to be used for arm templates, I think we should call them that.

The other reason we wanted to add an ArmTemplateParameters argument, was because the planned update to the subscription configuration added a field called the same. Adding the argument enables us to still be able to splat the subscription configuration object. It seems to me like for the short term, we should rename that field to AdditionalParameters and keep the parameter here as AdditionalParameters just so we don't have all these merging layers of multiple variables that all serve the same purpose. But I'd like to know how easy/hard it is to ascertain the dev impact of making a breaking change to this script (...maybe some telemetry would be nice, ha).

Reply:
"AdditionalParameters" is a common term in PowerShell modules, hence why I chose it. They weren't necessarily meant for any particular context. While the focus of the scripts was to deploy an ARM template, pre- and post-scripts can get those as well and do whatever a service needs. Since I strip the undefined parameters out when passed to ARM (deployment fails otherwise), these really were meant to be just extra parameters for whatever is needed.

The scripts were written so devs and the CI have a single, unified way of deploying resources. We use them a lot. For example, when I write new tests I need to record the sessions, so I run the script, run/record my tests (fix them or the code if needed), commit my changes, and remove the resource group. It's the same script and mostly the same processing for devs and CIs this way. Before, every repo and most services within those repos was solving this differently, so everyone had to figure it out. Clean-up also wasn't prevalent. The scripts unify the approach now.

Reply:
That's helpful context, thanks!

So aside from naming, the main constraint with needing a new parameter is wanting to switch pipelines away from having to define stringified powershell objects for additional parameters, and instead being able to use yaml objects instead. In order to this with backwards compat for various tests.yml files (which I plan to migrate), we need a separate parameter with an object type.

One limitation I didn't catch earlier is that azure pipelines cannot render yaml objects into strings. However they are planning to [add support for this](https://developercommunity.visualstudio.com/idea/880210/allow-type-casting-or-expression-function-from-yam.html). As a result, I'm thinking of editing this script a little to [build the object in the powershell script](https://github.com/microsoft/azure-pipelines-yaml/issues/427) string based on the new format. Then once we can to yaml->string conversions, both the parameter definition and backing code will be simplified.

Reply:
Ok I did make an assumption that "AdditionalParameters" was only for the ARM template parameters. I still like the idea of having explicit new parameters for specific scenarios like env variables and arm template parameters. If needed we can merge in AdditionalParameters in with the ArmTemplateParameters then filter before we deploy but overtime it would be good for folks to pass template parameters under the name ArmTemplateParameters so that it is clear what they are intended for.

Reply:
Actually, that approach might not be possible either, in which case we'll need to just use powershell object literals as strings for the moment :(

Reply:
PowerShell doesn't require stringified parameters. I use these all the time like so:
```powershell
New-TestResources.ps1 ... -AdditionalParameters @{enableHsm=$true}
```
Works fine. It's YAML that's requiring this, hence peoplehaving to pass `"true"` instead of `$true`.

Originally AdditionalParameters only goes to ARM template, yes. It wasn't necessarily meant to. Adding more parameters makes this harder to use - not easier. If no one else is using AdditionalParameters for anything other than ARM (easy to search quick), just use it for ARM parameters but keep the name.

My point was, though, I don't see why you need to split env vars and ARM template parameters. Why can't both be in the same hashtable?

Reply:
It's not powershell that's the problem, it's passing objects around to be rendered into string blocks that azure pipelines doesn't support. Pipelines cannot convert yaml to a string, hence why the existing templates pass around a stringified hash literal.

As far as I could tell across all repos it's only being used as arm template parameters, and even then only in three places for storage/cosmos (two in java, one in js).

Reply:
The way I see the issue now is we have to use a stringified object literal, which will either be a yaml string or a powershell object string (I favor the latter at the moment for consistency). The decisions to make are whether this gets defined as a platform variable or a pipeline variable (I favor the latter), and whether to introduce a new variable or not.

Reply:
Env vars also get defined at the template parameter level, not the platform variable level. The pipelines using custom arm template parameters today differ on a per-platform basis, so we can't merge the two.

Reply:
Per discussion, let's add `[hashtable] $ArmTemplateParameters` and `[hashtable] $EnvironmentVariables` separately. Docs should probably clearly state these should be hashtables of string, string

Be sure to merge `$ArmTemplateParameters` into `$AdditionalParameters` so the types therein aren't lost (e.g. `$true`).

Reply:
Updated. Since we're passing raw powershell text, we aren't constrained by string:string anymore.